### ISSUE TITLE
Random For Games

### ISSUE THREAD

**Issue by tomspilman** [mono/MonoGame#3772]
The main problem I've run into is that the `Random` on different versions of C# while using the same seeds do not result in the same sequence. This is mainly between Mono and .NET. 

We've also found the stock .NET `Random` is pretty slow. Even when doing integer randomization it under the hood uses doubles:

https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Random.cs#L168

Worse MS can't really change the existing `Random` to fix these problems as it could break existing apps.

Also things like not having a `Random.NextFloat(float min, float max)` makes things inconvenient for game developers.

This all suggests adding our own `Random` that is aimed at game development needs.  Mainly a fast replacement for `Random` with support for all the data types and ranges. We would also provide a helper template method to randomly shuffle elements in a list.

 - High efficiency.
 - Great distribution.
 - Deterministic when you need it.
 - Can be totally random when you don't.
 - Provide complete set of min/max range functions.
 - Full set of unit tests proving the implementation.

Thoughts and suggestions welcome.

**Comment by Shqrdx** [comment]
First we need contrive its name - I suggest GameRandom.

**Comment by tomspilman** [comment]
`Random` seems like a good name on its own.  Let the namespace do the rest of the work.

**Comment by Shqrdx** [comment]
So every time I write "using System" I also should write "using Random = Microsoft.XNA.Framework.Random" ?

**Comment by BaseCrosser** [comment]
I was searching for a good `Random` implementation because of the difference between Mono and .NET `Random` a while ago. 

I found `FastRandom` at <a href="http://www.codeproject.com/Articles/9187/A-fast-equivalent-for-System-Random">CodeProject</a>. Maybe this would be a good candidate. It has a article explaining what the difference is and why it is faster.

I hope this helps and thanks for the work you guys are doing on MonoGame.

**Comment by cra0zy** [comment]
> So every time I write "using System" I also should write "using Random = Microsoft.XNA.Framework.Random" ? Otherwise ambiguity happens.

I agree with you on that one.

> I suggest GameRandom.

Doesn't sound nice, how about "Rng"? Literally every programmer and gamer I've met is calling it that anyway...

**Comment by Shqrdx** [comment]
Rng ? from "Range" ? Sounds like C++. There are few other variations : MGRandom, RandomEx, Randomizer,

**Comment by cra0zy** [comment]
Rng means random number generator...

**Comment by cra0zy** [comment]
> literally every programmer

I stand corrected...

**Comment by Shqrdx** [comment]
Ah, my bad english.. (

**Comment by WardBenjamin** [comment]
I agree that the class should not be called `Random` because it conflicts with the `System.Random` class, but maybe the random class could be nested within MathHelper? It would be accessed as:

    using Microsoft.Xna.Framework;
    int randInt = MathHelper.Random.NextInt(); 
    int rantInt0 = MathHelper.Random.NextInt(23, 439);

This seems like a fairly good solution to me, and I'm interested in working on this. Suggestions?

**Comment by WardBenjamin** [comment]
For the actual algorithm, I am going to suggest either a simple xorshift or xorshift*. This is much simpler than PRNG algorithms such as MT, and actually passes more of the BigCrush tests than the MT (int the case of xorshift*). I am suggesting xorshift* rather than xorshift+ because, though xorshift+ is slightly faster, xorshift* is slightly more "secure" and passes more BigCrush tests. (I know that this isn't a huge consideration, but hey, it is a slight performance difference for a slight gain.)

**Comment by KonajuGames** [comment]
I agree it does appear ambiguous if our class was called `Random` as well, but the other side is how often do you reference the class type?  Once to create the instance and then the ambiguity is no longer a concern because from that point on you use the methods and properties of the instance.

> maybe the random class could be nested within MathHelper

Your example uses it like a static class, which it shouldn't be.  You create an instance with a known seed to get a deterministic sequence of "random" values or a default seed that is usually based on the system time in order to get a pseudo-random sequence.  You may even have several instances of the class to provide sequences of random numbers to different modules of your game where the act of getting a random number for one module should not affect the random sequence for another module.

**Comment by KonajuGames** [comment]
> For the actual algorithm, I am going to suggest either a simple xorshift or xorshift.

I think [xorshift](http://en.wikipedia.org/wiki/Xorshift) looks like a good candidate.  The particular variant is not a huge factor, I think.

**Comment by WardBenjamin** [comment]
Relating to your first comment, this could be used in an identical way to what you say, actually. An xorshift is a seeded random number generator, so it is easily viable (maybe 20 LOC) to have this structured in an instance rather than as a static class. I'll write something up during lunxh and submit it for review as a PR.

**Comment by theZMan** [comment]
In 10+ years I've never seen Random show up on a profiling report what are you guys doing with it :-)

The missing methods are trivially added with Extension methods also - I'm sure all of us have a library of those to share.

However the lack of consistency between Mono and Microsoft isn't good - I'm sort of amazed that Mono didn't try to match Microsoft but it probably predates any agreements on source. So we need something for that.

I'd use namespaces too - if I want to switch over it makes it easier to do IMO but I'm not too swayed on either way... MGRandom? Rng and abbreviations like that are generally frowned upon in .Net naming standards by Microsoft and ofter don't work across languages.

**Comment by KonajuGames** [comment]
Given that the variant is not a huge factor, based on the implementations on that Wikipedia page I linked to, I would much prefer the xorshift64 implementation that uses a single uint64 seed rather than the xorshift1024 that uses an array of 16 uint64 values and an int.

**Comment by WardBenjamin** [comment]
I can definitely do that. I like that there can be multiple seeds with the xorshift1024, which can be generated using a simple permutation of a single number, but I'll start at least with xorshift64 and we'll see from there.

**Comment by WardBenjamin** [comment]
Also, is there anywhere that I can find coding conventions for MonoGmae? I have only contributed a simple Protobuild fix previously.

**Comment by tgjones** [comment]
As a twist on the above suggestions, how about calling it `Random`, but making it a nested class inside `MathHelper`? That way you'd get the advantage of using a sensible name, but without the conflict with `System.Random`. You'd have to reference it by the scoped name, `MathHelper.Random`.

``` csharp
public static class MathHelper
{
    // ... other stuff

    public class Random
    {
		
    }
}
```

**Comment by WardBenjamin** [comment]
@tgjones Yeah, that is identical to what I suggested in my first comment, combined with KonajuGames's suggestion. I think it's the way to go.

**Comment by tgjones** [comment]
@WardBenjamin Ah, sorry - I misread your suggestion. In that case, ignore my previous comment :)

**Comment by WardBenjamin** [comment]
No problem :)

**Comment by KonajuGames** [comment]
> Also, is there anywhere that I can find coding conventions for MonoGame @KonajuGames ? I have only contributed a simple Protobuild fix previously.

See [Contributing to MonoGame](https://github.com/mono/MonoGame/wiki/Contributing-to-MonoGame).

**Comment by tomspilman** [comment]
Note we can totally crib ideas for the MonoGame implementation from the CoreClr code:

 https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Random.cs#L168

Also i came across this trick reading an [old Coding Horror post on how to randomly sort a deck of cards](http://blog.codinghorror.com/shuffling/):
```
var cards = Enumerable.Range(0, 51);
var shuffledcards = cards.OrderBy(a => Guid.NewGuid());
```
It reminds me that `Random(Environment.TickCount)`... which is what .NET does internally...  is actually a very poor.  It means the random is predictable which is bad for some sorts of games.  By default we should feed a `Guid` in as the seed to avoid that trap.

Also it would be nice to have a built in `Random.Shffule(IEnumerable)` in our implementation.

**Comment by WardBenjamin** [comment]
Currently, I have the random number gen seeded by a value xor'ed from System.DateTime.Now's elements, which seems to be satisfactory but can later be changed. I will submit a pull request as soon as I write a function to change the range of the ulong that is currently fed out of my xorshift function to int range. After that I will write functions for float, double, etc.

**Comment by KonajuGames** [comment]
There was a well-known online poker cheat where the shuffling of the cards was determined by a RNG seeded from the server time.  Since the server time was predictable, the cheaters could predict the cards that would come out next.

**Comment by WardBenjamin** [comment]
Huh, that's true. I will change it to use a GUID if everyone agrees. It's going to be kind of difficult to do that since it's xor'ed down to the millisecond, but that's not a problem to change.

**Comment by WardBenjamin** [comment]
Also, currently I am using a ulong as the seed. Is this acceptable? It is the easiest to consider when generating the numbers using bitwise operators, but could be changed with a little bit of typecasting.

**Comment by tomspilman** [comment]
> Also, currently I am using a ulong as the seed. Is this acceptable 

Sure... that is fine.  The `Guid` is 128bit... so you would xor the two 64bit halves together to pass as the default `ulong` seed on the default `Random` constructor.

For bonus points maybe make a `static long Random.GenerateSeed()` internally it would build the seed from a new `Guid`.  This way if a game wants a good seed and wants to sync it across machines in for example multiplayer... they can.  This hopefully dissuades people from using `DateTime.Now` junk.

Super bonus points.... add `long Random.Seed { get; set; }` and allow the user to get or reseed the existing random without recreation.

**Comment by danzel** [comment]
Should this be part of monogame, or an external library?

**Comment by tomspilman** [comment]
> Should this be part of monogame, or an external library?

Part of MonoGame for now.

If we ever figure out how to best split the math stuff... Vector2/3/4, BoundingSphere, Matrix, etc... from the core of MonoGame then we would move the random with that.

**Comment by Shqrdx** [comment]
> If we ever figure out how to best split the math stuff...

Oh, no please no, it would be very confusing..

**Comment by Helikin** [comment]
I just want to mention, SharpNEAT is not an option because it uses the GPL license (see http://sharpneat.sourceforge.net/licensing.htm).

A new RNG family and a useful comparison of RNGs (including xorshift) was recently published here: http://www.pcg-random.org/

**Comment by WardBenjamin** [comment]
That's pretty cool. I have a preliminary version of the Random class in MathHelper.cs in my branch: https://github.com/WardBenjamin/MonoGame/tree/Random with an xorshift private function and some type conversions that I would appreciate review of.

**Comment by KonajuGames** [comment]
It should be consistent in use of `long` and `ulong`.  The `GenerateSeed` method returns a `long`, the seed is stored as a `ulong`, the `Seed` property accepts and returns a `long`, and the ctor takes a `ulong`.  Choose one and use it consistently.

No need for throwing of exceptions and all the parameter validation in the `Remap` private method.  It's private and only called by methods that you write.  You should be able to ensure that it receives the expected parameters because you're writing the calls to it.

The multiplication in the XORShift64Star method probably needs to be unchecked because it will likely overflow.

Remapping a `ulong` to `byte`, `short`, `int`, etc will be just as effective with a simple cast rather than the potentially costly `Remap` method.  In the case of converting ulong to long, `unsafe { return *(long *)&uint64; }` (reinterpreting the bits as a different type of the same size) should do fine.

**Comment by Shqrdx** [comment]
The main differences between standart Random and new Random are 
-faster speed
-float generation

We could also add methods for generating random vectors,color etc...

**Comment by WardBenjamin** [comment]
@KonajuGames One of the issues with that is that ulong.MaxValue is interpreted as -1, and so on until the point where it wraps. Basically we would have to do some funky range calculations, since the ulong only wraps back around to be positive at `ulong.MaxValue - (ulong)long.MaxValue - 1UL`. `ulong.MaxValue - (ulong)long.MaxValue` is equivalent to long.MinValue, and subtracting one more is the same is long.MaxValue. I will write some code up in a sec. 

As a note, that unsafe code is unnecessary since it can be replaced with some calls to BitConverter methods, but is probably faster, so I'll use it.

**Comment by WardBenjamin** [comment]
Essentially, to map the ulong to the long correctly, the direct conversion needs to happen, then the values will be mapped by the following rules:

If the original value is less than `ulong.MaxValue - (ulong)long.MaxValue`, keep it.
If the original value is greater than or equal to `ulong.MaxValue - (ulong)long.MaxValue

**Comment by WardBenjamin** [comment]
That seems to work. Any suggestions for improvement other than the ones already given?

**Comment by tomspilman** [comment]
>  I have a preliminary version of the Random class in MathHelper.cs in my branch

 Looks like a good start.

First thing I would say would be we should probably move it into its own CS file regardless of how it is accessed.  It will be complex enough that it doesn't need to be mixed in with the rest of `MathHelpers`...  just make it a partial.

Also please prefix member fields with underscore... so `_seed` and `_lastValue`.

Will reseeding via restoring an old seed into `Random.Seed` work?  The `lastValue` member seems to suggest it needs to be restored as well?


> As a note, that unsafe code is unnecessary 

Unsafe code is fine... all our platforms support it.  Especially if it makes things faster.

**Comment by KonajuGames** [comment]
The seed value does not need to remain unchanged through the lifetime of
the object. It should be used as the lastValue as well.

**Comment by WardBenjamin** [comment]
I can definitely make those changes, super easy. As an additional note, I think my ulong to long function is suffering from an off by one, or else the remap function is. I'm going to do some more testing tonight.

**Comment by WardBenjamin** [comment]
Alright, I have completed the requested changes as well as implementing long -> other integer type conversion (floating point numbers are already complete).

**Comment by WardBenjamin** [comment]
Ran into a wall based on coding standards. Since CLS compliancy includes no exposed unsigned integer types, the ctor warns about not being CLS complaint according to the following rule:

(Copy-Pasta from http://www.devarticles.com/c/a/C-Sharp/Making-Your-Code-CLS-Compliant/)

    Unsigned types should not be part of the public interface of the class. What this means is public fields should not have unsigned types like uint or ulong, public methods should not return unsigned types, parameters passed to public function should not have unsigned types. However unsigned types can be part of private members. 

Should I change this or leave it as-is?

**Comment by KonajuGames** [comment]
There are already several other places in the code that use the `[ClsCompliant(false)]` attribute.

**Comment by WardBenjamin** [comment]
Alright, I will mark it as so then.

**Comment by paewie** [comment]
@tomspilman i didn't really catch your intention of avoiding the DateTime stuff.
Okay, actually I do catch it because I know a bit about cryptology, but using a GUID is just as predictable as using the current time. I don't recall where I read this, but based on a single GUID the next about 128 can be predicted, which makes it an "irrelevant" choice for CSPRNGs.
So there remains one question to me: Do you want a CSPRNG or just a default PRNG? Or even a real RNG?

Searching around a bit and having your suggestion of using the half of the GUID to seed the xorshift64 I found this http://blogs.msdn.com/b/oldnewthing/archive/2008/06/27/8659071.aspx

**Comment by theZMan** [comment]
Unless you are using Random for security or running a for profit online casino predictability really shouldn't be a big deal for any of our games... seems to me we are heading right down Over Design Alley on this one ;-)

**Comment by paewie** [comment]
@theZMan I know. But @tomspilman actually stated
"It reminds me that Random(Environment.TickCount)... which is what .NET does internally... is actually a very poor. It means the random is easily predictable which is bad for some sorts of games. By default we should feed a Guid in as the seed to avoid that trap."
speaking of predictability. This confused me because used in Games there should be no need to make it truly random - as you said. Also he said that using a GUID would avoid this predictability. So he either wanted to get a CSPRNG - overkill for games - or it doesn't / shouldn't matter anyways.

**Comment by WardBenjamin** [comment]
@paewie We actually do take the whole GUID as a seed by Xor'ing it together in the GenerateSeed() method, fyi.

**Comment by tomspilman** [comment]
> but using a GUID is just as predictable as using the current time.

But it is not.  I just now generated a random seed at 6:19pm CST.  Please tell me what it was?

If the seed was based on time only you could maybe take some guesses and probably get it.  If it was based on GUID it is much harder to guess...  you would need access to details of my machine you do not have.

This is my only point...  ticks or time is very poor and GUID is much better.

**Comment by theZMan** [comment]
Turns out System.Random seeds from Environment.TickCount  https://msdn.microsoft.com/en-us/library/system.environment.tickcount(v=vs.110).aspx so I'd need to know when you turned your machine on and be within a few milliseconds of my guess. 

Even if it used DateTime.Ticks that's a nanosecond based timer so 6:19pm gives me about a billion possible seeds to chose from depending on how accurate your clock is. They are sequential so I suppose I could try them all eventually.

**Comment by tomspilman** [comment]
> Turns out System.Random seeds from Environment.TickCount 
> DateTime.Ticks that's a nanosecond based timer 

 Both correct, but doesn't change that GUID is still more unique than plain time.

**Comment by paewie** [comment]
@timspilman maybe my unlucky chose of words made me sound a bit rude. I'm sorry for that.
As said: I was just confused since you seemed to want a "secure" seed. A Guid isn't much of that. Of course it's more than a tick timer, but nothing relevant for cryptography. Therefore I wanted to point that out. Nothing more (since I don't catch the purpose for a "secure" rng in games anyway).

If you want to be exact this Guid or Environment.TickCount seeds both can be brute-forced, a tick counter more easily than a Guid. That way or another it would be easier to break the algorithm in another way - like guessed plain text since games have very limited possible plain states that can be randomized/encrypted. Talking in terms of cryptology both are poor, but (both) suitable for a simple prng. But if you want a Guid, then take it ;)

**Comment by paewie** [comment]
@WardBenjamin oh, okay. I didn't realize that. I just read the comments on this post and made a quick look at your commits, but nothing very thoroughly... Since XOR is a strange thing that could even "harden" the seed a little. I can't tell for sure.

**Comment by tomspilman** [comment]
>  I was just confused since you seemed to want a "secure" seed.

Yeah... i'm not suggesting that at all.  If we wanted that we would use the .NET crypto classes like `RNGCryptoServiceProvider`, but I worry that won't work correctly on anything but desktop .NET.

**Comment by paewie** [comment]
Probably... I don't know how .Net or Mono handle encryption. If they match at all: Just on Windows. Unix is using different native libraries than mac than win etc.
But okay. Since you don't want to provide encryption I just misunderstood your avoidance of the time-thing :)

**Comment by WardBenjamin** [comment]
@tomspilman My pull request for this is complete as far as the RNG portion of it is concerned, I will open another afterwards after this current one is reviewed and approved. Will you or @KonajuGames review it? 

Another topic: what utility functions do you actually want me to include? You have previously mentioned a shuffle for IEnumerables, which is really easy with a GUID, but what else do you want?

**Comment by tomspilman** [comment]
> Will you or @KonajuGames review it? 

We will both review it.

> what utility functions do you actually want me to include?

 Not sure off the top of my head...  but what we do really need is aunit test proving the new class works as expected.

> You have previously mentioned a shuffle for IEnumerables, which is really easy with a GUID,

We shouldn't use GUIDs for that... we should use the random to do it.

**Comment by WardBenjamin** [comment]
I have updated the xml docs a little bit, now onto this "unit test". Do we have any documentation on how to write one of these? I have never seen/written one before. @tomspilman

**Comment by tomspilman** [comment]
> Do we have any documentation on how to write one of these? 

The `MonoGame.Tests.XXX` project is including the solution.  There is no documentation on how to write these, but you can use existing tests as a sort of guide:

https://github.com/mono/MonoGame/blob/develop/Test/Framework/Vector3Test.cs

I would create a new GameRandomTest.cs file and add in tests to verify that the random works as expected.  

If you run the EXE it will execute all the tests which makes things hard to test at times.  If you have something like Reflector or [NUnit Runner](https://visualstudiogallery.msdn.microsoft.com/6ab922d0-21c0-4f06-ab5f-4ecd1fe7175d) plugging installed into VS then you can right click and execute just a single test at a time.

**Comment by WardBenjamin** [comment]
When building the NUnit tests solution, it completely fails with a ton of null-exceptions, argument exceptions, and such. Some of these seem to be intentional, such as the argument exceptions. Am I correct in ignoring those?

**Comment by tomspilman** [comment]
> I have just regenerated the project and solution files and am using the WindowsGL solution for testing

Don't use the WindowsGL unit tests... they are still pretty broken.  Use the regular Windows DirectX platform for writing and testing your unit tests.

**Comment by WardBenjamin** [comment]
Alright, I can do that.

**Comment by teryror** [comment]
As for utility functions, in addition to the already mentioned NextFloat(min, max) and Shuffle(IEnumerable), there's a couple useful things I can think of:

* Sampling from non-uniform distributions, especially the normal distribution. I'd imagine that poisson distributions would come in handy with particle effects and the like as well.
* Sampling from arbitrary, discrete distributions is a pretty common problem and can be done very efficiently with the alias method (explained very well here: http://www.keithschwarz.com/darts-dice-coins/). Very useful for drop tables, though we'd need an additional class AliasTable. Might be considered overkill.
* Simpler stuff like T Pick<T>(params T[] source) and sampling data with and without replacement (the former is easy, the latter has been described by Knuth: http://rosettacode.org/wiki/Knuth%27s_algorithm_S).

Finally, I would second @Helikin's suggestion of using PCG instead of xorshift as the internal engine, because (a) it has some additional utility available, like rollback and jump ahead, and more importantly (b) you can spawn multiple instances that actually produce unique sequences, rather than just start at different seed-locations in the same sequence.
This is valuable for anyone using multiple instances, since seeding with GUIDs doesn't guarantee unique states: For any 64-bit state, there are 2^64 128-bit numbers that produce the same state. Not all 128-bit numbers are valid GUIDs, but there will probably still be collisions.

I understand @WardBenjamin's been working on the xorshift generator for a while, but I have a working, unit tested implementation of PCG and some of the utility functions I mentioned on my PC, and I could integrate it into the MonoGame source over the next couple days. Opinions?

Even if you reject the PCG idea, I'd still be willing to help with utility functions, of course.

**Comment by WardBenjamin** [comment]
I would be completely fine as long as you kept the API/documentation and such as it is, or similar. Actually, something that could save a lot of effort is if @teryror just pulled my branch in my fork down and used the API that is already there, just integrating his PCG implementation. I am having a bit of trouble coming up with an extensive test since it sometimes takes millions of iterations to get the maximum and minimum possible variations of the NextInt methods, for example, and there are a bunch of these methods to test.

If your PCG implementation is already unit-testable and such I suggest that you pull down this fork and modify it to preserve the documentation and such, even if you just obliterate the rest of the code. I'd like my work to stand if possible :smile:

I'm going to start working on some of the utility functions assuming you will do that, I can take care of at least the shuffle function and do some brainstorming for others.

**Comment by teryror** [comment]
I already forked your branch and am currently cloning it. Still new to github and VCS in general, so no need to risk messing up your branch.

As for unit tests, the ones I have are based on Microsoft.VisualStudio.TestTools.UnitTesting, rather than NUnit, so I'll have to do some work porting them over. Also, it's not like I was particularly thorough testing the generator itself (private project and all that), I'll have to add some more to prove the correctness of the implementation.
I do have lots of tests for edge cases in my utility functions, though. However, the trick I used there relied on my implementation being a bit over-engineered: I currently have an interface defining a sample method, that a concrete generator (like PCG) would implement. Utility functions are extension methods for that interface, so I could easily make a mock-generator that consistently produced the edge cases.
We'll have to think of something else to test them (probably something about calculating a seed to instantaneously generate the numbers we want, then using that as a magic seed in the test. Although that adds dependency on the sampling method to the unit test for utility functions...).

Quick API question: currently, there's a member _seed that holds the original seed, and _state that holds the current state. Getting the Seed property returns _seed, while setting it sets both. Should we keep it that way, or reduce it to just _state and have a property for that?

**Comment by WardBenjamin** [comment]
> Should we keep it that way, or reduce it to just _state and have a property for that?

Honestly whatever works for you is fine. We ended up deciding on this so that the Seed actually represented the initial value of the RNG, while maintaining a separate state internally so that we could use it in the xorshift calculation.

**Comment by teryror** [comment]
> so that the Seed actually represented the initial value of the RNG, while maintaining a separate state internally

Makes sense. No point in touching that then.

**Comment by teryror** [comment]
Okay, so I just synced a couple of commits to https://github.com/teryror/MonoGame/blob/Random/MonoGame.Framework/Random.cs, which is pretty much done in my point of view (Your mileage may vary, especially concerning https://github.com/teryror/MonoGame/commit/a2d59f2b04033ac7bcc425507e4e88cd894ae303 (the NextInt32() "alias" for Next() which is just plain old code duplication, imo) and https://github.com/teryror/MonoGame/commit/c9ebc6c4c17f31939d8db1253515ccd7a0fe7b4e.

Also, I changed the output range of NextFloat and NextDouble to [0, 1), to match the behaviour of System.Random.

I didn't add any unit tests yet, because I still have to figure out how to run them; there are no *.nunit files to open with the NUnit runner, and the project doesn't compile because it can't find the [ClsCompliant] attribute, which should be in mscorlib, as per https://msdn.microsoft.com/en-us/library/system.clscompliantattribute%28v=vs.110%29.aspx. Weird.

**Comment by WardBenjamin** [comment]
@teryror I have updated my branch with your changes as well as updating the pull request. We'll have to find out @tomspilman 's opinion about the Nextint32() alias removal as he is the one who wanted that in place IIRC. Other than that, it looks very good. Nice job.

I think the unit tests are the only thing left to complete, again you'd have to check with Tom about the ClsCompliant thing, I pulled it down and mscorlib seems to be linked, so I have no idea what the problem is.

**Comment by teryror** [comment]
Unit tests are up and running, though [this](http://i.imgur.com/uvnj7Pn.png) just doesn't look right. Then again, that's 127 *errors*, it's not like these tests *failed* per se. I didn't muck with any code outside Random.cs and [RandomTest.cs](https://github.com/teryror/MonoGame/blob/Random/Test/Framework/RandomTest.cs), so those must have been broken before?

I still have to figure out how to sensibly force edge cases on the existing utility functions (I take it you tried brute forcing it, @WardBenjamin?).

Before we can close this issue, we should also talk about more utility functions once more, no? IIRC, @tomspilman wanted this to be a drop in replacement for System.Random, so we'll need, at minimum, NextBytes(byte[]).

It's like 2am over here though, so I second letting this rest until the higher ups speak :smile:

**Comment by WardBenjamin** [comment]
@teryror Yeah, I tried brute forcing it and it took a really long time, especially for the functions that can return a larger domain (the NextInt64 methods especially). I don't think that's the best option here, but I don't really have any better ideas. I can definitely write NextBytes right now, I need a short break from working on my game anyhow.

It's not a bad idea to wait for a response, I have no idea where to proceed from here really.

**Comment by WardBenjamin** [comment]
An update for anyone who has not seen the PR: at the moment there is a set of unit tests and the PR is ready to be merged on @tomspilman's approval. The tests could be improved but they are usable as-is.

**Comment by WardBenjamin** [comment]
Are there any updates on the approval status of this PR? 

CC @KonajuGames @tomspilman

**Comment by WardBenjamin** [comment]
How do I get some attention back on this? It still needs review by @tomspilman or @KonajuGames or somebody.

Ideas?

CC @Shqrdx, @cra0zy, others.

**Comment by BrainSlugs83** [comment]
Kind of catching up on this now -- I have a bunch of utility extension methods for System.Random that I use, that I would be willing to contribute, everything from Vector2 / Vector2d, double, float, Color, all kinds of stuff.

One thing that I would recommend, just from reading the thread:

If the Seed property really just holds the initial seed, I would recommend changing it to be "InitialSeed" -- and then exposing the current seed / internal state as a property "CurrentSeed".

Another recommendation, kind of torn, might be a bad idea, but it seems like if the set's on the seed property have side effects (i.e. it changes other properties as well, etc.) that they should just be read only properties, and that reseeding should be accomplished via a method (i.e. "public void Reseed(...)").

**Comment by WardBenjamin** [comment]
I kind of like Seed vs State. It's made clear by the documentation what happens when you use the different ones. The method makes it even more ambiguous what happens, especially since the Seed and State properties are technically methods themselves.

TL; DR I think we shouldn't change this; it's fairly clear and we agreed on these semantics previously.

Feel free to contribute the utility methods though! :smiley:

**Comment by tomspilman** [comment]
@WardBenjamin - Sorry... things have been crazy busy here.  I'll get back to this soon.

**Comment by WardBenjamin** [comment]
Awesome, good to see some progress on this!

**Comment by BrainSlugs83** [comment]
Oh, I hadn't realized there was a whole separate discussion thread on the pull request.  D'oh.

**Comment by WardBenjamin** [comment]
So..... @tomspilman It's been a bit over a week. Any updates?

Just worrying that this issue and PR are getting buried :smile:

**Comment by craftworkgames** [comment]
Wow this is a really long thread. I'm surprised there can be so much discussion about an implementation of Random.

Anyway, I just wanted to pipe in here and say that I remember seeing a pretty good implementation of a fast random in the Mercury Particle Engine.
https://github.com/Matthew-Davey/mercury-particle-engine/blob/develop/Mercury.ParticleEngine.Core/FastRand.cs

**Comment by WardBenjamin** [comment]
@craftworkgames That fast random is essentially what I had originally, if you check out my first few commits. We switched to a much better (randomness, other attributes) RNG contributed by @teryror at a slight speed loss (has it even been measured), eventually.

**Comment by craftworkgames** [comment]
@WardBenjamin Good to know.

**Comment by SimonDarksideJ** [comment]
Implemented through
https://github.com/MonoGame/MonoGame/pull/8861
